tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/Metric.java

```diff
@@ -0,0 +1,196 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.framework.metrics;
+
+import org.tensorflow.Operand;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.types.family.TNumber;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Base class for Metrics
+ *
+ * @param <U> The data type for the metric values
+ * @param <T> The data type for the metric result
+ */
+public abstract class Metric<U extends TNumber, T extends TNumber> {
+
+  /** The TensorFlow Ops */
+  private final Ops tf;
+
+  private final long seed;
+
+  /** The name for this metric. Defaults to {@link Class#getSimpleName()}. */
+  private final String name;
+
+  /**
+   * Creates a Metric with a name of {@link Class#getSimpleName()}
+   *
+   * @param tf the TensorFlow Ops
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   */
+  protected Metric(Ops tf, long seed) {
```

Comment:
That's a valid point, though for now most of the library relies exclusively on compile-time safety type checks, I think it is fare for this new code to follow the actual paradigm, that could be revisited later if needed.

Reply:
Runtime type checking wasn't my main motivation. We already have situations where we store a runtime type because we need it later, at a point in the code where we don't have a tensor in hand to pull it from. Here's an example at org/tensorflow/framework/metrics/impl/MeanMetricWrapper.java:98, the `getResultType()` calls:
~~~ Java
    Operand<T> tLabels = CastHelper.cast(getTF(), labels, getResultType());
    Operand<T> tPredictions = CastHelper.cast(getTF(), predictions, getResultType());

    Operand<T> losses = loss.call(tLabels, tPredictions);
~~~
If we had an `Operand<T>` in hand we could pull the runtime type from there, but in this situation we don't. So we take the runtime type as a constructor parameter:
~~~ Java
public class MeanMetricWrapper<U extends TNumber, T extends TNumber> extends Mean<U, T> {
   . . .
  protected MeanMetricWrapper(Ops tf, String name, long seed, Class<T> type) {
    super(tf, name, seed, type);
  }
~~~
But suppose we didn't need a stored runtime class when we originally implemented this class, but then eventually made changes that required it? (Or suppose we need the runtime type for `U`!) It would be too late to require the runtime class in every constructor. Instead, we'd have to choose between unpleasant alternatives:
* Add a new constructor that takes the runtime class, and make it a runtime error to use the new functionality if you didn't use that constructor.
* Punt on adding the new functionality to the existing class hierarchy, perhaps instead creating a `MeanMetricWrapperV2` and then also forking subclasses as needed.
* Or make a breaking change by adding a runtime type parameter to existing constructors (which in this case would break a ton of subclass APIs).

I wonder whether we should make it a consistent pattern that classes parameterized by tensor types take the runtime types in their constructors?

Reply:
Ok I see what you mean. If we decide to take that direction though, like you said, we would need to go through all classes we have created so far and see if they would be good candidates take a type parameter in their constructor as well to be consistent.

Therefore I would not prevent merging this PR in its actual state and maybe suggest to create an issue to continue that discussion and do a more complete analysis, wdyt?

Reply:
👍 nothing to do in this PR